internal/kibana/slo.go

```diff
@@ -0,0 +1,506 @@
+
+	default:
+		return models.Slo{}, diag.Errorf("unknown indicator type %s", indicatorType)
+	}
+
+	timeWindow := slo.TimeWindow{
+		Type:     d.Get("time_window.0.type").(string),
+		Duration: d.Get("time_window.0.duration").(string),
+	}
+
+	objective := slo.Objective{
+		Target:          d.Get("objective.0.target").(float64),
+		TimesliceTarget: getOrNilFloat("objective.0.timeslice_target", d),
+		TimesliceWindow: getOrNilString("objective.0.timeslice_window", d),
+	}
+
+	settings := slo.Settings{
+		SyncDelay: getOrNilString("settings.0.sync_delay", d),
+		Frequency: getOrNilString("settings.0.frequency", d),
+	}
+
+	budgetingMethod := slo.BudgetingMethod(d.Get("budgeting_method").(string))
+
+	slo := models.Slo{
+		Name:            d.Get("name").(string),
+		Description:     d.Get("description").(string),
+		Indicator:       indicator,
+		TimeWindow:      timeWindow,
+		BudgetingMethod: budgetingMethod,
+		Objective:       objective,
+		Settings:        &settings,
+		SpaceID:         d.Get("space_id").(string),
+		GroupBy:         getOrNilString("group_by", d),
+	}
+
+	return slo, diags
+}
+
+func resourceSloCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClient(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+
+	slo, diags := getSloFromResourceData(d)
+	if diags.HasError() {
+		return diags
+	}
+
+	res, diags := kibana.CreateSlo(ctx, client, slo)
+
+	if diags.HasError() {
+		return diags
+	}
+
+	id := &clients.CompositeId{ClusterId: slo.SpaceID, ResourceId: res.ID}
+	d.SetId(id.String())
+
+	return resourceSloRead(ctx, d, meta)
+}
+
+func resourceSloUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClient(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+
+	slo, diags := getSloFromResourceData(d)
+	if diags.HasError() {
+		return diags
+	}
+
+	compId, diags := clients.CompositeIdFromStr(d.Id())
+	if diags.HasError() {
+		return diags
+	}
+	slo.ID = compId.ResourceId
+
+	res, diags := kibana.UpdateSlo(ctx, client, slo)
+
+	if diags.HasError() {
+		return diags
+	}
+
+	id := &clients.CompositeId{ClusterId: slo.SpaceID, ResourceId: res.ID}
+	d.SetId(id.String())
+
+	return resourceSloRead(ctx, d, meta)
+}
+
+func resourceSloRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClient(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+	compId, diags := clients.CompositeIdFromStr(d.Id())
+	if diags.HasError() {
+		return diags
+	}
+	id := compId.ResourceId
+	spaceId := compId.ClusterId
+
+	s, diags := kibana.GetSlo(ctx, client, id, spaceId)
+	if s == nil && diags == nil {
+		d.SetId("")
+		return nil
+	}
+	if diags.HasError() {
+		return diags
+	}
+
+	indicator := []interface{}{}
+	if s.Indicator.IndicatorPropertiesApmAvailability != nil {
```

Comment:
Does it make sense to replace `if` with `switch`?

Reply:
I'm not sure how that's possible? I'm not really switching on a value here.

Reply:
Sure, if you like. Seems to fit the style of the rest of the project better. Will just have a little `if` at the end to catch the case where an indicator isn't set for some reason.